### PR TITLE
Add IBM Cloud IPI Deployment support

### PR DESCRIPTION
**Pull request by clacroix12**
Jira: https://issues.redhat.com/browse/OCSQE-768

Signed-off-by: Coady LaCroix <clacroix@redhat.com>

### CODE REVIEW

**Review by ebenahar — DISMISSED** [review]

**Review by ebenahar — DISMISSED** [review]

**Comment by clacroix12** [comment]
Latest testing is showing that the cluster is up and running, however I'm seeing an error during the noobaa health check. Looking into that issue, but I believe this is ready for review.

Screen Shot 2022-09-08 at 10 52 47 AM

**Comment by clacroix12** [comment]
ODF deployment is now succeeding with the fix from #6492:

```
23:00:37 - MainThread - tests.ecosystem.deployment.test_deployment - INFO  - Verifying ceph health after deployment
23:00:37 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: oc wait --for condition=ready pod -l app=rook-ceph-tools -n openshift-storage --timeout=120s
23:00:38 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: oc -n openshift-storage get pod -l 'app=rook-ceph-tools' -o jsonpath='{.items[0].metadata.name}'
23:00:38 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: oc -n openshift-storage exec rook-ceph-tools-6c87ffd67-rztpb -- ceph health
23:00:39 - MainThread - ocs_ci.utility.utils - INFO  - Ceph cluster health is HEALTH_OK.
```

The last issue I'm seeing is with teardown. Often I see Timeouts from the destroy cluster command, however the assets all end up destroyed in IBM Cloud. I think I can bump the timeout slightly to accommodate this, but logically this PR is ready for reviews and is verified.

**Review by dahorak — APPROVED** [review]

**Review by petr-balogh — DISMISSED** [review]
I would like to see also some documentation for this addition for IPI deployment here:
https://github.com/red-hat-storage/ocs-ci/blob/master/docs/supported_platforms.md#ibm-cloud - also can be done in follow up PR

File: ocs_ci/deployment/ibmcloud.py [hunk 1]
```diff
@@ -111,3 +121,219 @@ def check_cluster_existence(self, cluster_name_prefix):
...
+    def configure_cloud_credential_operator():
+        """
+        Extract and Prepare the CCO utility (ccoctl) binary. This utility
+        allows us to create and manage cloud credentials from outside of
+        the cluster while in manual mode.
+
+        """
+        # retrieve ccoctl binary from https://mirror.openshift.com
+        version = (
+            "4.11.0"  # For now this is stable, if necessary make this configurable
```
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/ibmcloud.py:320]: I assume we want this tied to the OCP version, I can include it there.

File: conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+---
+DEPLOYMENT:
+  openshift_install_timeout: 4800
+  allow_lower_instance_requirements: false
+ENV_DATA:
```
[the remark this answers is not in the document]
Comment (clacroix12) [reply conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml:5]: Yes, we do. We need the api_key and account_id:

```
---
AUTH:
  ibmcloud:
    api_key: "<api_key>"
    account_id: "<account_id>"


```

I currently have this in a separately created credential file that I'm passing via  `--ocsci-conf` and planned on adding it to jenkins as a `CREDENTIALS_CONF`. I can add a similar comment that explains these values are necessary.

File: ocs_ci/deployment/ibmcloud.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/ibmcloud.py]: @petr-balogh I'm not sure of the exact mapping between OCP versions and what is supported for ccoctl so for now I'm placing the version in the default config so we can override it from OCP configs once we know.

File: conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (clacroix12) [reply conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml:5]: Added a similar section to this config.

File: conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  platform: 'ibm_cloud'
```
Comment (shylesh) [inline conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml:6]: @clacroix12 , do you feel like having same naming convention for platform key as well. I mean, without '_'  i.e  'ibmcloud' ?

File: ocs_ci/deployment/ibmcloud.py [hunk 4]
```diff
@@ -111,3 +121,217 @@ def check_cluster_existence(self, cluster_name_prefix):
...
+            log_level (str): log level openshift-installer (default: DEBUG)
+
+        """
+        self.export_api_key()
+        logger.info("Destroying the IBM Cloud cluster")
+        super(IBMCloudIPI, self).destroy_cluster(log_level)
+        self.delete_service_id()
+        resource_group = self.get_resource_group()
+        self.delete_volumes(resource_group)
+        time.sleep(30)  # Volumes take time to fully delete
```
Comment (shylesh) [inline ocs_ci/deployment/ibmcloud.py:178]: @clacroix12 , can we also check if entries are actually got deleted ?

File: ocs_ci/deployment/ibmcloud.py [hunk 5]
```diff
@@ -111,3 +121,217 @@ def check_cluster_existence(self, cluster_name_prefix):
...
+        def _delete_group():
+            cmd = f"ibmcloud resource group-delete {resource_group} -f"
+            exec_cmd(cmd)
+
+        if resource_group:
+            logger.info(f"Deleting resource group: {resource_group}")
+            _delete_group()
+
+    @staticmethod
+    def check_cluster_existence(cluster_name_prefix):
```
Comment (shylesh) [inline ocs_ci/deployment/ibmcloud.py:285]: @clacroix12 , can this be a base class method ? Or is it different from managed cluster?

File: ocs_ci/deployment/ibmcloud.py [hunk 6]
```diff
@@ -111,3 +121,217 @@ def check_cluster_existence(self, cluster_name_prefix):
...
+        self.delete_service_id()
+        resource_group = self.get_resource_group()
+        self.delete_volumes(resource_group)
+        time.sleep(30)  # Volumes take time to fully delete
+        self.delete_resource_group(resource_group)
+
+    def manually_create_iam_for_vpc(self):
+        """
+        Manually specify the IAM secrets for the cloud provider
+        """
```
Comment (shylesh) [inline ocs_ci/deployment/ibmcloud.py:184]: Empty line before the end of docstring is not consistent.

**Review by shylesh — COMMENTED** [review]
some minor comments, otherwise LGTM.

File: conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml (hunk 3 above)
Comment (clacroix12) [reply conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml:6]: `ibm_cloud` is already one of our platform values that we use throughout the framework, this wasn't decided with the addition of IPI. See: https://github.com/red-hat-storage/ocs-ci/blob/master/ocs_ci/ocs/constants.py#L843

File: ocs_ci/deployment/ibmcloud.py (hunk 5 above)
Comment (clacroix12) [reply ocs_ci/deployment/ibmcloud.py:285]: The base class is `CloudDeploymentBase` and as such this is too specific to be included in that class. This has IBM specifics. You can see this also implemented in the `IBMCloud` class. I *could* create another base class these both pull from but in the interest of keeping things much simpler I chose to just implement this method in my new class.

File: ocs_ci/deployment/ibmcloud.py (hunk 6 above)
Comment (clacroix12) [reply ocs_ci/deployment/ibmcloud.py:184]: Generally we include the empty line if it is more than a one liner and omit it if the docstring is just one line (description). There are some inconsistencies with other docstrings though so I will address those.

File: conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (petr-balogh) [reply conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml:5]: Will you also update the doc in follow up PR? [supported_platforms.md#ibm-cloud](https://github.com/red-hat-storage/ocs-ci/blob/master/docs/supported_platforms.md#ibm-cloud)

File: ocs_ci/deployment/ibmcloud.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (petr-balogh) [reply ocs_ci/deployment/ibmcloud.py]: Can we raise this question somewhere?

**Review by petr-balogh — APPROVED** [review]

File: conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (clacroix12) [reply conf/deployment/ibmcloud/ipi_3az_rhcos_3m_3w.yaml:5]: Sorry I missed the original message on the review. I can also update the `supported_platforms` doc.

File: ocs_ci/deployment/ibmcloud.py (hunk 4 above)
Comment (clacroix12) [reply ocs_ci/deployment/ibmcloud.py]: @shylesh Updated `delete_volumes` to verify that volumes are deleted before moving on. Still need to test this, I'm deploying a cluster right now so I can do so.

File: ocs_ci/deployment/ibmcloud.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/ibmcloud.py]: @petr-balogh Yeah I can reach out and see if there is such a mapping / supported versions list.

**Review by shylesh — APPROVED** [review]
LGTM

File: ocs_ci/deployment/ibmcloud.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/ibmcloud.py]: Nothing official but the suggestion I got was to use the binary that corresponds to the same OCP version, at least the major.minor version. So for 4.11 our default value makes sense, with 4.12 we can set it to 4.12.0 (or whatever patch version they have available).

File: ocs_ci/deployment/ibmcloud.py (hunk 4 above)
Comment (clacroix12) [reply ocs_ci/deployment/ibmcloud.py]: Some output from my testing:

```
23:51:21 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: ibmcloud resource groups --output json
23:51:22 - MainThread - ocs_ci.deployment.ibmcloud - INFO  - Retrieving cluster resource group
23:51:22 - MainThread - ocs_ci.deployment.ibmcloud - INFO  - Found resource group: clacroix-ibm-rv5gk
23:51:22 - MainThread - ocs_ci.deployment.ibmcloud - INFO  - Deleting volumes
23:51:22 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: ibmcloud is vols --resource-group-name clacroix-ibm-rv5gk --output json
23:51:29 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: ibmcloud is vold -f r006-7f8970e2-140c-4536-8d6a-4edcb39d3fcd r006-1a1603d0-b79e-46e5-95a4-567f182d4e70 r006-8e3fa3f1-52a3-46b4-a803-38e1b939ae1d r006-eee5a5d1-e66f-4816-8545-753f172acc6b r006-48e38184-7b51-4c33-aeb4-f380a106e814 r006-010e006f-8b88-4869-9db4-ab0b62fdb860
23:51:30 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: ibmcloud is vols --resource-group-name clacroix-ibm-rv5gk --output json
23:51:31 - MainThread - ocs_ci.utility.retry - WARNING  - Volumes still exist in resource group, Retrying in 10 seconds...
23:51:41 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: ibmcloud is vols --resource-group-name clacroix-ibm-rv5gk --output json
23:51:42 - MainThread - ocs_ci.utility.retry - WARNING  - Volumes still exist in resource group, Retrying in 10 seconds...
23:51:52 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: ibmcloud is vols --resource-group-name clacroix-ibm-rv5gk --output json
23:51:53 - MainThread - ocs_ci.utility.retry - WARNING  - Volumes still exist in resource group, Retrying in 10 seconds...
23:52:03 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: ibmcloud is vols --resource-group-name clacroix-ibm-rv5gk --output json
23:52:04 - MainThread - ocs_ci.deployment.ibmcloud - INFO  - Deleting resource group: clacroix-ibm-rv5gk
23:52:04 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: ibmcloud resource group-delete clacroix-ibm-rv5gk -f
```

**Review by petr-balogh — DISMISSED** [review]

File: ocs_ci/deployment/ibmcloud.py [hunk 7]
```diff
@@ -111,3 +120,234 @@ def check_cluster_existence(self, cluster_name_prefix):
...
+        exec_cmd(cmd)
+        delete_file(tarball)
+
+    @staticmethod
+    def export_api_key():
+        """
+        Exports the IBM CLoud API key as an environment variable.
+        """
+        logger.info("Exporting IC_API_KEY environment variable")
+        api_key = config.AUTH["ibm_cloud"]["api_key"]
```
Comment (petr-balogh) [inline ocs_ci/deployment/ibmcloud.py:352]: we are using `ibmcloud` key

Comment (clacroix12) [reply ocs_ci/deployment/ibmcloud.py]: fixed

**Comment by clacroix12** [comment]
One more quick fix to pull the `bin_dir` from the config instead of constructing it from the constants. This way the `ccoctl` binary is placed in the correct location for our jenkins pipelines as well.

**Review by petr-balogh — DISMISSED** [review]

**Review by petr-balogh — APPROVED** [review]

**Comment by clacroix12** [comment]
Rebased on master and ran PR verification again to ensure this code is compatible with the master branch. Verified and ready for reviews/approvals.

**Review by petr-balogh — APPROVED** [review]

**Pull request merged**